### PR TITLE
Add transforms for all SSL models

### PR DESCRIPTION
**Pull request by Ra1Nik**
Add transforms for all SSL models with tests

### CODE REVIEW

File: lightly/transforms/models/moco_transform.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+from lightly.transforms.models.simclr_transform import SimCLRTransform
+from lightly.transforms.utils import IMAGENET_NORMALIZE
+from typing import Optional, Tuple, Union
+
+
+class MoCoTransform(SimCLRTransform):
+    """Implements the transformations for MoCo v1.
+
+    For MoCo v2, simply use the SimCLR settings.
```
Comment (guarin) [inline lightly/transforms/models/moco_transform.py:9]: I think we can drop the comment about MoCo v2 and just add a "new" `MoCoV2Transform` at the end of the file by setting `MoCoV2Transform = SimCLRTransform` with a comment above that says that moco v2 transform settings are identical to simclr.

File: lightly/transforms/models/simclr_transform.py [hunk 2]
```diff
@@ -58,6 +58,7 @@ def __init__(
         min_scale: float = 0.15,
         random_gray_scale: float = 0.2,
         gaussian_blur: float = 0.5,
+        kernel_size: Optional[float] = None,
```
Comment (guarin) [inline lightly/transforms/models/simclr_transform.py:61]: I noticed that `input_size` should be `224` in `SimCLRTransform` and `SimCLRViewTransform`

File: lightly/transforms/models/msn_transform.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        cj_strength: float = 1.0,
+        gaussian_blur: float = 0.5,
+        kernel_size: Optional[float] = None,
+        sigmas: Tuple[float, float] = (0.2, 2),
+        random_gray_scale: float = 0.2,
+        hf_prob: float = 0.5,
+        vf_prob: float = 0.0,
+        normalize: dict = IMAGENET_NORMALIZE,
+    ):
+        transform = MSNViewTransform(
```
Comment (guarin) [inline lightly/transforms/models/msn_transform.py:72]:
```suggestion
        random_view_transform = MSNViewTransform(
```

File: lightly/transforms/models/msn_transform.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            cj_strength=cj_strength,
+            gaussian_blur=gaussian_blur,
+            kernel_size=kernel_size,
+            sigmas=sigmas,
+            random_gray_scale=random_gray_scale,
+            hf_prob=hf_prob,
+            vf_prob=vf_prob,
+            normalize=normalize,
+        )
+        focal_transform = MSNViewTransform(
```
Comment (guarin) [inline lightly/transforms/models/msn_transform.py:85]:
```suggestion
        focal_view_transform = MSNViewTransform(
```

File: lightly/transforms/models/pirl_transform.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        # Cropping and normalisation for non-transformed image
+        no_augment = T.Compose(
+            [
+                T.RandomResizedCrop(size=input_size, scale=(min_scale, 1.0)),
+                T.ToTensor(),
+                T.Normalize(mean=normalize["mean"], std=normalize["std"]),
+            ]
+        )
+        jigsaw = Jigsaw(
```
Comment (guarin) [inline lightly/transforms/models/pirl_transform.py:84]: Could you move the definition of `transform` just above `jigsaw = Jigsaw`? Then the code is more grouped. First we define the `no_augment` transform and then the `jigsaw transform`.

File: lightly/transforms/models/smog_transform.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        random_gray_scale:
+            Probability of conversion to grayscale.
+        normalize:
+            Dictionary with 'mean' and 'std' for torchvision.transforms.Normalize.
+
+    """
+
+    def __init__(
+        self,
+        crop_sizes: Tuple[int] = (224, 96),
```
Comment (guarin) [inline lightly/transforms/models/smog_transform.py:46]:
```suggestion
        crop_sizes: Tuple[int, int] = (224, 96),
```

Similar for the tuples below.

File: lightly/transforms/models/smog_transform.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        crop_counts: Tuple[int] = (4, 4),
+        crop_min_scales: Tuple[float] = (0.2, 0.05),
+        crop_max_scales: Tuple[float] = (1.0, 0.2),
+        gaussian_blur_probs: Tuple[float] = (0.5, 0.1),
+        gaussian_blur_kernel_sizes: Optional[Tuple[float]] = (None, None),
```
Comment (guarin) [inline lightly/transforms/models/smog_transform.py:51]:
```suggestion
        gaussian_blur_kernel_sizes: Tuple[Optional[float], Optional[float]] = (None, None),
```

File: lightly/transforms/models/smog_transform.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        gaussian_blur_sigmas: Tuple[float, float] = (0.2, 2),
+        solarize_probs: Tuple[float] = (0.0, 0.2),
```
Comment (guarin) [inline lightly/transforms/models/smog_transform.py:53]: -> `Tuple[float, float]`

File: lightly/transforms/models/smog_transform.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+
+class SmoGViewTransform:
+    def __init__(
+        self,
+        crop_size: int = 224,
+        crop_min_scale: float = 0.2,
+        crop_max_scale: float = 1.0,
+        gaussian_blur_prob: float = 0.5,
+        kernel_size: Optional[Tuple[float]] = None,
```
Comment (guarin) [inline lightly/transforms/models/smog_transform.py:94]:
```suggestion
        kernel_size: Optional[float] = None,
```

File: lightly/transforms/models/vicreg_transform.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            rr_degrees=rr_degrees,
+            normalize=normalize,
+        )
+        super().__init__(transforms=[view_transform, view_transform])
+
+
+class VICRegViewTransform:
+    def __init__(
+        self,
+        input_size: int = 64,
```
Comment (guarin) [inline lightly/transforms/models/vicreg_transform.py:106]: I guess this should be `224`

File: lightly/transforms/models/vicreg_transform.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        cj_prob: float = 0.8,
+        cj_bright: float = 0.7,
+        cj_contrast: float = 0.7,
+        cj_sat: float = 0.7,
+        cj_hue: float = 0.2,
+        min_scale: float = 0.15,
```
Comment (guarin) [inline lightly/transforms/models/vicreg_transform.py:112]: The default values don't match the values in `VicRegCollate`

File: lightly/transforms/models/vicregl_transform.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        global_gaussian_blur_prob:
+            Probability of Gaussian blur for the global crop category.
+        local_gaussian_blur_prob:
+            Probability of Gaussian blur for the local crop category.
+        global_gaussian_blur_kernel_size:
+            Will be deprecated in favor of `global_gaussian_blur_sigmas` argument. If set, the old behavior applies and `global_gaussian_blur_sigmas` is ignored.
+            Used to calculate sigma of gaussian blur with global_gaussian_blur_kernel_size * input_size. Applied to global crop category.
+        local_gaussian_blur_kernel_size:
+            Will be deprecated in favor of `local_gaussian_blur_sigmas` argument. If set, the old behavior applies and `local_gaussian_blur_sigmas` is ignored.
+            Used to calculate sigma of gaussian blur with local_gaussian_blur_kernel_size * input_size. Applied to local crop category.
```
Comment (guarin) [inline lightly/transforms/models/vicregl_transform.py:37]: Could you add some line breaks in here to make the lines a bit shorter?

File: lightly/transforms/image_grid_transform.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+class ImageGridTransform:
+    """Transforms an image into multiple views.
+
+    Args:
+        transforms:
+            A sequence of transforms. Every transform creates a new view and its
+            correspondent local grid. Used for VICRegL. This argument in particular is
+            a list of tuples comprehending the cropping and the corrisponding grid
+            transformation and the non geometrical composition of transforms.
```
Comment (guarin) [inline lightly/transforms/image_grid_transform.py:15]: Could you add some more info on the expected input? From VIGRegLTransform I gathered that the input should be `[(image_grid_transform, view_transform), (image_grid_transform, view_transform), ...]`

File: lightly/transforms/image_grid_transform.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+                [3, local_crop_size, local_crop_size],
+                [global_grid_size, global_grid_size, 2],
+                [local_grid_size, local_grid_size, 2]
+            ]
+
+        """
+        views, grids = [], []
+        for transform in self.transforms:
+            view, grid = transform[0].forward(image)
+            views.append(transform[1](view))
```
Comment (guarin) [inline lightly/transforms/image_grid_transform.py:45]:
```suggestion
        for image_grid_transform, view_transform in self.transforms:
            view, grid = image_grid_transform(image)
            views.append(video_transform(view))
```

File: tests/transforms/models/test_dino_transform.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    assert output.shape == (3, 32, 32)
+
+
+def test_multi_view_on_pil_image():
+    multi_view_transform = DINOTransform(global_crop_size=32, local_crop_size=8)
+    sample = Image.new("RGB", (100, 100))
+    output = multi_view_transform(sample)
+    assert len(output) == 8
+    for output_img in output:
+        assert output_img.shape == (3, 32, 32) or output_img.shape == (3, 8, 8)
```
Comment (guarin) [inline tests/transforms/models/test_dino_transform.py:19]: I would prefer to make the assertions explicit:
```
# global views
assert all(out.shape == (3, 32, 32) for out in output[:2])
# local views
assert all(out.shape == (3, 8, 8) for out in output[2:])

File: tests/transforms/models/test_msn_transform.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    assert output.shape == (3, 32, 32)
+
+
+def test_multi_view_on_pil_image():
+    multi_view_transform = MSNTransform(random_size=32, focal_size=8)
+    sample = Image.new("RGB", (100, 100))
+    output = multi_view_transform(sample)
+    assert len(output) == 12
+    for output_img in output:
+        assert output_img.shape == (3, 32, 32) or output_img.shape == (3, 8, 8)
```
Comment (guarin) [inline tests/transforms/models/test_msn_transform.py:19]: Please assert shapes exactly.

File: tests/transforms/models/test_smog_transform.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    assert output.shape == (3, 32, 32)
+
+
+def test_multi_view_on_pil_image():
+    multi_view_transform = SMoGTransform(crop_sizes=(32, 8))
+    sample = Image.new("RGB", (100, 100))
+    output = multi_view_transform(sample)
+    assert len(output) == 8
+    for output_img in output:
+        assert output_img.shape == (3, 32, 32) or output_img.shape == (3, 8, 8)
```
Comment (guarin) [inline tests/transforms/models/test_smog_transform.py:19]: Please assert shapes exactly.

File: tests/transforms/models/test_swav_transform.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+    assert output.shape == (3, 100, 100)
+
+
+def test_multi_view_on_pil_image():
+    multi_view_transform = SwaVTransform(crop_sizes=(32, 8))
+    sample = Image.new("RGB", (100, 100))
+    output = multi_view_transform(sample)
+    assert len(output) == 8
+    for output_img in output:
+        assert output_img.shape == (3, 32, 32) or output_img.shape == (3, 8, 8)
```
Comment (guarin) [inline tests/transforms/models/test_swav_transform.py:19]: Please assert shapes exactly.

File: lightly/transforms/mae_transform.py [hunk 19]
```diff
@@ -59,10 +44,11 @@ def __call__(self, image: Union[Tensor, Image]) -> Tensor:
...
+            image: 
+                The input image to apply the transforms to.
 
         Returns:
-            Image (Tensor): The transformed image.
+            The transformed image.
 
         """
-        return self.transform(image)
+        return [self.transform(image)]
```
Comment (guarin) [inline lightly/transforms/mae_transform.py:54]: We should just return the single image and not inherit from `MultiViewTransform` as MAE doesn't really create multiple views.

File: lightly/transforms/image_grid_transform.py [hunk 20]
```diff
@@ -13,6 +13,12 @@ class ImageGridTransform:
             correspondent local grid. Used for VICRegL. This argument in particular is
             a list of tuples comprehending the cropping and the corrisponding grid
             transformation and the non geometrical composition of transforms.
+            
+            For VICRegL it comprehends:
+            [
+                (global_image_grid_transform, global_view_transform), 
+                (local_image_grid_transform, local_view_transform)
+            ]
```
Comment (guarin) [inline lightly/transforms/image_grid_transform.py:21]: I would propose the following docstring:
```
    """Transforms an image into multiple views and grids.

    Used for VICRegL.

    Attributes:
        transforms:
            A sequence of (image_grid_transform, view_transform) tuples. 
            The image_grid_transform creates a new view and grid from the image.
            The view_transform further augments the view. Every transform tuple
            is applied once to the image, creating len(transforms) views and
            grids.
    """
```

File: lightly/transforms/vicregl_transform.py [hunk 21]
```diff
@@ -61,8 +67,8 @@ def __init__(
         self,
         global_crop_size: int = 224,
         local_crop_size: int = 96,
-        global_crop_scale: Tuple[int] = (0.2, 1.0),
-        local_crop_scale: Tuple[int] = (0.05, 0.2),
+        global_crop_scale: Tuple[float, float] = (0.2, 1.0),
+        local_crop_scale: Tuple[float, float] = (0.05, 0.2),
```
Comment (guarin) [inline lightly/transforms/vicregl_transform.py:71]: Below in the code you set `self.transforms = ...`, I think you shouldn't do this and just use `transforms = ..` instead as you pass the variable to `super().__init__()` afterwards. The `ImageGridTransform` will then set `self.transforms`

Comment (guarin) [reply lightly/transforms/vicregl_transform.py:71]: Same for the `VICRegViewTransform`

**Review by guarin — CHANGES_REQUESTED** [review]
A few last comments and then this is good to merge :)

File: lightly/transforms/vicregl_transform.py (hunk 21 above)
Comment (Ra1Nik) [reply lightly/transforms/vicregl_transform.py:71]: I can't see the issue in the `ViewTransform`

Comment (guarin) [reply lightly/transforms/vicregl_transform.py:71]: Ah sorry, you are right, the `ViewTransform` should store `self.transform`. All good!

**Review by guarin — APPROVED** [review]

**Pull request merged**